File: src/main/java/com/perfscope/ui/CallTreeCell.java
package com.perfscope.ui;

import com.perfscope.model.CallTreeData;
import com.perfscope.util.Duration;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.control.TreeCell;
import javafx.scene.layout.StackPane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Rectangle;

public class CallTreeCell extends TreeCell<CallTreeData> {
    private final StackPane stack = new StackPane();
    private final Rectangle timeBar = new Rectangle();
    private final Label label = new Label();
    
    public CallTreeCell() {
        timeBar.setHeight(20);
        timeBar.setFill(Color.web("#4285F4", 0.3));
        timeBar.setArcWidth(5);
        timeBar.setArcHeight(5);
        
        stack.getChildren().addAll(timeBar, label);
        stack.setAlignment(Pos.CENTER_LEFT);
        stack.setPadding(new Insets(2, 5, 2, 0));
        
        StackPane.setAlignment(timeBar, Pos.CENTER_LEFT);
        timeBar.setTranslateX(0);
        
        label.setStyle("-fx-font-weight: normal;");
    }
    
    @Override
    protected void updateItem(CallTreeData item, boolean empty) {
        super.updateItem(item, empty);
        
        if (empty || item == null) {
            setText(null);
            setGraphic(null);
        } else {
            String labelText;
            if (item.getTotalTime() != null) {
                labelText = String.format("%s [%s]", item.getName(), Duration.ofNanos(item.getTotalTime()));
            } else {
                labelText = item.getName();
            }

            label.setText(labelText);
            
            if (item.getTimeNanos() != null && item.getTimeNanos() > 0) {
                double ratio = item.getTimeRatio();
                double maxWidth = getTreeView().getWidth() - (getTreeView().getRoot().getChildren().size() > 0 ? 
                                                             getTreeView().getRoot().getChildren().get(0).getGraphic() != null ? 
                                                             40 : 20 : 20);
                timeBar.setWidth(Math.max(1, ratio * maxWidth));
                timeBar.setVisible(true);
            } else {
                timeBar.setVisible(false);
            }
            
            setGraphic(stack);
        }
    }
} 

File: src/main/kotlin/com/perfscope/view/CallTreeCell.kt
package com.perfscope.view

import com.perfscope.model.Call
import com.perfscope.util.Duration.Companion.ofNanos
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.control.TreeCell
import javafx.scene.layout.StackPane
import javafx.scene.paint.Color
import javafx.scene.shape.Rectangle
import javafx.scene.text.Text
import javafx.scene.text.TextFlow
import kotlin.math.max

class CallTreeCell : TreeCell<Call?>() {
    private val stack = StackPane()
    private val timeBar = Rectangle()
    private val textFlow = TextFlow()

    init {
        timeBar.height = 20.0
        timeBar.fill = Color.web("#4285F4", 0.3)
        timeBar.arcWidth = 5.0
        timeBar.arcHeight = 5.0

        stack.children.addAll(timeBar, textFlow)
        stack.alignment = Pos.CENTER_LEFT
        stack.padding = Insets(2.0, 5.0, 2.0, 0.0)

        StackPane.setAlignment(timeBar, Pos.CENTER_LEFT)
        timeBar.translateX = 0.0
    }

    override fun updateItem(item: Call?, empty: Boolean) {
        super.updateItem(item, empty)
        textFlow.children.clear()

        if (empty || item == null) {
            text = null
            graphic = null
        } else {
            val labelText = Text()
            if (item.totalTime != null) {
                labelText.text = String.format("%s [%s]", item.name, ofNanos(item.totalTime))
            } else {
                labelText.text = item.name
            }
            labelText.styleClass += "perfscope-call-tree"

            textFlow.children += labelText

            if (item.timeNanos != null && item.timeNanos > 0) {
                val ratio: Double = item.timeRatio
                val maxWidth: Double = treeView.width - (if (treeView.getRoot().getChildren().isNotEmpty()) if
                        (treeView.getRoot().getChildren()[0].graphic != null) 40 else 20 else 20)
                timeBar.width = max(1.0, ratio * maxWidth)
                timeBar.isVisible = true
            } else {
                timeBar.isVisible = false
            }

            graphic = stack
        }
    }
}